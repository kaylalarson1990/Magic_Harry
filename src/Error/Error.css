.page-not-found {
    font-size: 48px;
    color: #BE0125;
}