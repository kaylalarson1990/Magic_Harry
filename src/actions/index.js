export const setCharacters = characters => ({
  type: "SET_CHARACTERS",
  characters
});

export const setHouses = houses => ({
  type: "SET_HOUSES",
  houses
});

export const setSpells = spells => ({
  type: "SET_SPELLS",
  spells
});
