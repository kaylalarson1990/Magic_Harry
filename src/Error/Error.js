import React from "react";
import "./Error.css";
import sortinghat from "../images/sortinghat.png";

const Error = () => {
  return (
    <section>
      <h2 className="page-not-found">
        404 ERROR: PAGE NOT FOUND{" "}
        <img src={sortinghat} alt="wand" className="sorting-hat" />
      </h2>
    </section>
  );
};

export default Error;
