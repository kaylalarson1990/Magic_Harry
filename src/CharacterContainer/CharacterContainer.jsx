import React, { Component } from "react";
import { connect } from "react-redux";
import "./CharacterContainer.scss";
import { harryPotterCharacters } from "../apiCalls/apiCalls";
import Characters from "../Characters/Characters.jsx";
import { setCharacters } from "../actions/index";
import PropTypes from "prop-types";

class CharacterContainer extends Component {
  constructor() {
    super();
    this.state = {
      error: ""
    };
  }
  componentDidMount() {
    harryPotterCharacters()
      .then(data => data)
      .then(characters => this.props.setCharacters(characters))
      .catch(this.setState({ error: "Error fetching wizard data" }));
  }

  render() {
    const { characters } = this.props;
    const displayCharacters = characters.map(character => (
      <Characters
        key={character.id}
        name={character.name}
        house={character.house}
        role={character.role}
        school={character.school}
        species={character.species}
      />
    ));
    return (
      <>
        <article>
          <h2 className="title">Harry Potter Characters</h2>
          <input class="search__input" type="text" placeholder="Search" />
        </article>
        <section className="characters">{displayCharacters}</section>
      </>
    );
  }
}

const mapStateToProps = state => ({
  characters: state.characters
});

const mapDispatchToProps = dispatch => ({
  setCharacters: characters => dispatch(setCharacters(characters))
});

CharacterContainer.propTypes = {
  characters: PropTypes.object,
  setCharacters: PropTypes.array
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CharacterContainer);
