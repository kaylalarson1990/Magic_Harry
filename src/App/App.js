import React, { Component } from "react";
import { connect } from "react-redux";
import "./App.css";
import Header from "../Header/Header";
import Error from "../Error/Error.js";
import Favorites from "../Favorites/Favorites";
import CharacterContainer from "../CharacterContainer/CharacterContainer";
import { MainContainer } from "../MainContainer/MainContainer";
import HouseContainer from "../HouseContainer/HouseContainer";
import SpellContainer from "../SpellContainer/SpellContainer";
import {
  harryPotterSpells,
  harryPotterHouses,
  harryPotterCharacters
} from "../apiCalls/apiCalls";
import { setSpells, setHouses, setCharacters } from "../actions/index";
import { Route, Switch } from "react-router-dom";

export class App extends Component {
  constructor() {
    super();
    this.state = {
      favorites: []
    };
  }

  componentDidMount() {
    harryPotterSpells()
      .then(data => data)
      .then(spells => this.props.setSpells(spells))
      .catch(this.setState({ error: "Error fetching wizard data" }));
    harryPotterHouses()
      .then(data => data)
      .then(houses => this.props.setHouses(houses))
      .catch(this.setState({ error: "Error fetching wizard data" }));
    harryPotterCharacters()
      .then(data => data)
      .then(characters => this.props.setCharacters(characters))
      .catch(this.setState({ error: "Error fetching wizard data" }));
  }

  favoriteCard = async id => {
    const favorites = this.state.favorites;
    const favoritedCard = await [
      ...this.props.spells,
      ...this.props.houses,
      ...this.props.characters
    ].find(card => card.id === id);

    if (!favorites.includes(favoritedCard)) {
      favoritedCard.favorite = !favoritedCard.favorite;
      this.setState({
        favorites: [...favorites, favoritedCard]
      });
    } else {
      favoritedCard.favorite = !favoritedCard.favorite;
      this.setState({
        favorites: favorites.filter(favorite => favorite.id !== id)
      });
    }
  };

  homePage = () => (
    <>
      <MainContainer />
    </>
  );

  render() {
    return (
      <div className="App">
        <Header />
        <Switch>
          <Route exact path="/" component={() => this.homePage()} />
          <Route
            exact
            path="/characters"
            render={() => (
              <CharacterContainer
                data={this.props.characters}
                favoriteCard={this.favoriteCard}
              />
            )}
          />
          <Route
            exact
            path="/houses"
            component={() => (
              <HouseContainer
                data={this.props.houses}
                favoriteCard={this.favoriteCard}
              />
            )}
          />
          <Route
            exact
            path="/spells"
            component={() => (
              <SpellContainer
                data={this.props.spells}
                favoriteCard={this.favoriteCard}
              />
            )}
          />
          <Route
            exact
            path="/favorites"
            render={() => (
              <Favorites
                favorites={this.state.favorites}
                favoriteCard={this.favoriteCard}
                onLoad={this.getFromStorage}
              />
            )}
          />
          <Route render={Error} />
        </Switch>
      </div>
    );
  }
}

export const mapStateToProps = state => ({
  spells: state.spells,
  houses: state.houses,
  characters: state.characters
});

export const mapDispatchToProps = dispatch => ({
  setSpells: spells => dispatch(setSpells(spells)),
  setHouses: houses => dispatch(setHouses(houses)),
  setCharacters: characters => dispatch(setCharacters(characters))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(App);
