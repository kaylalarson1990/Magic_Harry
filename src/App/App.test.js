import React from "react";
import { shallow, mount } from "enzyme";
import { App } from "./App";
import Error from "../Error/Error";
import { CharacterContainer } from "../CharacterContainer/CharacterContainer";
import { MemoryRouter } from "react-router";
import { Route } from "react-router-dom";
import { setCharacters, setHouses, setSpells } from "../actions/index";
import { mapStateToProps, mapDispatchToProps } from "./App";

describe("App", () => {
  let wrapper, instance, mockFunc;

  beforeEach(() => {
    wrapper = shallow(<App />);
    instance = wrapper.instance();
    mockFunc = jest.fn();
  });

  it("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render home page", () => {
    expect(instance.homePage()).toMatchSnapshot();
  });

  it("should be able to save to state", () => {
    const mockData = {
      error: "Error fetching wizard data",
      favorites: []
    };
    expect(wrapper.state()).toEqual(mockData);
  });

  it("should save to localStorage", () => {
    jest.spyOn(window.localStorage.__proto__, "setItem");
    window.localStorage.__proto__.setItem = jest.fn();

    window.localStorage.__proto__.setItem();
    expect(localStorage.setItem).toHaveBeenCalled();
  });

  it("should redirect to 404", () => {
    const wrapper = shallow(
      <MemoryRouter initialEntries={["/random"]}>
        <App />
      </MemoryRouter>
    );

    expect(wrapper.find(instance.homePage())).toHaveLength(0);
  });

  it("should redirect CharacterContainer to 404", () => {
    const mockProps = [
      {
        name: "unknown"
      }
    ];
    const wrapper = shallow(
      <MemoryRouter initialEntries={["/random"]} initialIndex={0}>
        <App />
      </MemoryRouter>
    );
    expect(
      wrapper.find(
        <CharacterContainer data={mockProps} favoriteCard={mockFunc} />
      )
    ).toHaveLength(0);
  });
});

describe("mapDispatchToProps", () => {
  it("calls dispatch with a setCharacters action", () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = setCharacters("Kayla");
    const mappedDispatch = mapDispatchToProps(mockDispatch);
    mappedDispatch.setCharacters("Kayla");
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it("calls dispatch with a setHouses action", () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = setHouses("Slytherin");
    const mappedDispatch = mapDispatchToProps(mockDispatch);
    mappedDispatch.setHouses("Slytherin");
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it("calls dispatch with a setSpells action", () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = setSpells("Accio");
    const mappedDispatch = mapDispatchToProps(mockDispatch);
    mappedDispatch.setSpells("Accio");
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });
});

describe("mapStateToProps", () => {
  it("should return an object with a characters array", () => {
    const mockState = {
      characters: "Kayla",
      type: "SET_CHARACTERS"
    };
    const expected = {
      characters: "Kayla"
    };
    const mappedProps = mapStateToProps(mockState);
    expect(mappedProps).toEqual(expected);
  });

  it("should return an object with a houses array", () => {
    const mockState = {
      houses: "Slytherin",
      type: "SET_HOUSES"
    };
    const expected = {
      houses: "Slytherin"
    };
    const mappedProps = mapStateToProps(mockState);
    expect(mappedProps).toEqual(expected);
  });

  it("should return an object with a spells array", () => {
    const mockState = {
      spells: "Accio",
      type: "SET_SPELS"
    };
    const expected = {
      spells: "Accio"
    };
    const mappedProps = mapStateToProps(mockState);
    expect(mappedProps).toEqual(expected);
  });
});
