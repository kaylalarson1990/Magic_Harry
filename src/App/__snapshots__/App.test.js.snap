// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<div
  className="App"
>
  <Connect(Header) />
  <Switch>
    <Route
      component={[Function]}
      exact={true}
      path="/"
    />
    <Route
      exact={true}
      path="/characters"
      render={[Function]}
    />
    <Route
      component={[Function]}
      exact={true}
      path="/houses"
    />
    <Route
      component={[Function]}
      exact={true}
      path="/spells"
    />
    <Route
      exact={true}
      path="/favorites"
      render={[Function]}
    />
    <Route
      render={[Function]}
    />
  </Switch>
</div>
`;

exports[`App should render home page 1`] = `
<React.Fragment>
  <MainContainer />
</React.Fragment>
`;
